pkg/apis/akash.network/v2beta1/types.go

```diff
@@ -0,0 +1,456 @@
+	HTTPOptions ManifestServiceExposeHTTPOptions `json:"http_options,omitempty"`
+}
+
+type ManifestServiceExposeHTTPOptions struct {
+	MaxBodySize uint32   `json:"max_body_size,omitempty"`
+	ReadTimeout uint32   `json:"read_timeout,omitempty"`
+	SendTimeout uint32   `json:"send_timeout,omitempty"`
+	NextTries   uint32   `json:"next_tries,omitempty"`
+	NextTimeout uint32   `json:"next_timeout,omitempty"`
+	NextCases   []string `json:"next_cases,omitempty"`
+}
+
+func (mse ManifestServiceExpose) toAkash() (manifest.ServiceExpose, error) {
+	proto, err := manifest.ParseServiceProtocol(mse.Proto)
+	if err != nil {
+		fmt.Printf("foobar: %q\n", mse.Proto)
+		return manifest.ServiceExpose{}, err
+	}
+	return manifest.ServiceExpose{
+		Port:         mse.Port,
+		ExternalPort: mse.ExternalPort,
+		Proto:        proto,
+		Service:      mse.Service,
+		Global:       mse.Global,
+		Hosts:        mse.Hosts,
+	}, nil
+}
+
+func manifestServiceExposeFromAkash(amse manifest.ServiceExpose) ManifestServiceExpose {
+	return ManifestServiceExpose{
+		Port:         amse.Port,
+		ExternalPort: amse.ExternalPort,
+		Proto:        amse.Proto.ToString(),
+		Service:      amse.Service,
+		Global:       amse.Global,
+		Hosts:        amse.Hosts,
+		HTTPOptions: ManifestServiceExposeHTTPOptions{
+			MaxBodySize: amse.HTTPOptions.MaxBodySize,
+			ReadTimeout: amse.HTTPOptions.ReadTimeout,
+			SendTimeout: amse.HTTPOptions.SendTimeout,
+			NextTries:   amse.HTTPOptions.NextTries,
+			NextTimeout: amse.HTTPOptions.NextTimeout,
+			NextCases:   amse.HTTPOptions.NextCases,
+		},
+	}
+}
+
+type ManifestServiceStorage struct {
+	Name string `json:"name"`
+	Size string `json:"size"`
+}
+
+// ResourceUnits stores cpu, memory and storage details
+type ResourceUnits struct {
+	CPU     uint32                   `json:"cpu,omitempty"`
+	Memory  string                   `json:"memory,omitempty"`
+	Storage []ManifestServiceStorage `json:"storage,omitempty"`
+}
+
+func (ru ResourceUnits) toAkash() (types.ResourceUnits, error) {
+	memory, err := strconv.ParseUint(ru.Memory, 10, 64)
+	if err != nil {
+		return types.ResourceUnits{}, err
+	}
+
+	storage := make([]types.Storage, 0, len(ru.Storage))
+	for _, st := range ru.Storage {
+		size, err := strconv.ParseUint(st.Size, 10, 64)
+		if err != nil {
+			return types.ResourceUnits{}, err
+		}
+
+		storage = append(storage, types.Storage{
+			Name:     st.Name,
+			Quantity: types.NewResourceValue(size),
+		})
+	}
+
+	return types.ResourceUnits{
+		CPU: &types.CPU{
+			Units: types.NewResourceValue(uint64(ru.CPU)),
+		},
+		Memory: &types.Memory{
+			Quantity: types.NewResourceValue(memory),
+		},
+		Storage: storage,
+	}, nil
+}
+
+func resourceUnitsFromAkash(aru types.ResourceUnits) (ResourceUnits, error) {
+	res := ResourceUnits{}
+	if aru.CPU != nil {
+		if aru.CPU.Units.Value() > math.MaxUint32 {
+			return ResourceUnits{}, errors.Errorf("k8s api: cpu units value overflows uint32")
```

Comment:
This isn't a format string, just use `errors.New` to a create a single error object